---
title: "First step towards quantitative predictions of planktonic functional diversity from physico-chemical data"
output: html_notebook
---
Emile Faure
Last modification : 26/03/2020
Please note that what we describe as Protein Functional Clusters in our article's main text is here called CC, for connected components.

```{r}

# PACKAGES IMPORTATION
library(igraph)
library(ggplot2)
theme_set(theme_minimal()) 
library(vegan)
library(gridExtra)
library(caret)
library(glmnet)
library(reshape2)
library(TeachingDemos)
library(plm)
library(car)
library(ggrepel)
library(randomForest)
library(plyr)
library(dplyr)
library(doParallel)
library(foreach)
library(maps)
source("http://peterhaschke.com/Code/multiplot.R")

```

We already have genes quantifications, genes annotations, and sets of nodes and edges based on genes similarities and coverage. We first need to build the sequence similarity network, and derive abundance tables from it.

```{r}
############### BUILDING AN ABUNDANCE TABLE #################
#  Emile Faure, 2019

###### Data Importation ######
quant_genes = read.table("/media/DATA/Emile/TARA_Genomics/Salmon_Quant_Output/quantif_genes", header = T)
v_80 = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Proka_MAGs_Diamond_out_1e-3_no-selfhits_pident80_cov80.blastp.vertices", h=T, sep=";", quote=NULL)
e_80 = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Proka_MAGs_Diamond_out_1e-3_no-selfhits_pident80_cov80.blastp.edges", h=T, sep=";")

# Let's reduce the quant_genes file to the 757,457 genes found in the SSN
length(which(quant_genes$Name %in% v_80$name)) # 757,457

quant_genes = quant_genes[which(quant_genes$Name %in% v_80$name),]

# Now we need to sum abundances of the 163 metagenomes to match the 93 samples
samples = read.table("/home/faure/Documents/PhD/MAGs_Project/Sample_ERRID_NbSeqPairs.csv", header = T, sep = ";")
summary(samples)

abund_genes = as.data.frame(quant_genes$Name)
names(abund_genes) = "Sample_ID"
for (i in c(1:nrow(samples))) {
  abund_genes[,i+1] = rep(0,nrow(abund_genes)) # Initialize column
  errid = which(names(quant_genes) %in% unlist(samples[i,2:5])) # get column numbers in quant_genes corresponding to focal sample
  for (j in c(1:length(errid))) {
   abund_genes[,i+1] = abund_genes[,i+1] + quant_genes[,errid[j]]   # add each column
  }
  names(abund_genes)[i+1] = as.character(samples[i,1]) # Rename column
  abund_genes[,i+1] = abund_genes[,i+1]/samples[i,6]*1000000000 # Normalize by Nb QC seqpairs
}

row.names(abund_genes) = abund_genes[,1]
abund_genes = abund_genes[,-1]

# We note that the 49th sample, "MED_23_05M" has an outlier value
# It's corresponding to ERR are ERR315858 & ERR315861
# The gene is TARA_MED_MAG_00097_000000000001_1, it is not outlying because of a mistake, 
# as it is very much present in the salmon data already (196248 reads mapped to it)
# Moreover, it is similarly expressed in the 2 duplicates ERR315858 & ERR315861
# The sequence is quite long, and appears in only one PFC

#write.table(abund_genes, "/home/faure/Documents/PhD/MAGs_Project/TabAbund_Genes_Mags")
#read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/TabAbund_Genes_Mags")

######## Abundance table for CCs

g_80 = graph.data.frame(e_80, directed=F, vertices=v_80)
#rm(quant_genes,e_80,samples)

# We conserve all CCs, using min vertices = 2
comp = decompose.graph(g_80, min.vertices = 2)

vertex_in_cc = c()
corres_cc = c()
j=1
for (i in comp) {
  vertex=V(i)$name
  vertex_in_cc = c(vertex_in_cc,vertex)
  corres_cc = c(corres_cc,rep(j,length(vertex)))
  j=j+1
}

length(vertex_in_cc)
length(corres_cc)

corres_cc = paste(rep("CC_",length(corres_cc)),corres_cc, sep = "")

abund_cc = data.frame(CC_ID = corres_cc, Genes = vertex_in_cc)

abund_cc = merge(x=abund_cc,y=abund_genes,by.x = "Genes",by.y="row.names")
row.names(abund_cc)=abund_cc$Genes
abund_cc = abund_cc[,-1]
head(abund_cc)

#write.table(abund_cc, "/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Tables_CC_2/TabAbund_CC_NOsum_CC2")
#abund_cc = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Tables_CC_2/TabAbund_CC_NOsum_CC2")

# Now a version with mean abundance by CC
abund_cc_mean = aggregate(abund_cc[,-1], list(abund_cc$CC_ID), mean)
row.names(abund_cc_mean)=abund_cc_mean$Group.1
abund_cc_mean = abund_cc_mean[,-1]

#write.table(abund_cc_mean, "/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Tables_CC_2/TabAbund_CC_mean_CC2")

#Empty environment before next section
rm(list=ls())
```

Now we can explore the size, the functional and taxonomical homogeneity of the connected components.

```{r}
############### Investigating protein functional clusters composition #################
#  Emile Faure, 2019

###### Data Importation ######
abund_cc = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Tables_CC_2/TabAbund_CC_NOsum_CC2", header = T)
abund_cc$protID = row.names(abund_cc)
abund_cc = abund_cc[,c(1,95)]

# Here we will only need the cc id column
annots_egg = read.table("/media/DATA/Emile/TARA_Genomics/Annotations/Allprot_eggNOG_Functional_description", header = F, sep="\t", quote="",  na.strings = "")
names(annots_egg)=c("protID", "Annot")
KOprot = read.table("/media/DATA/Emile/TARA_Genomics/Annotations_KOFamScan/KOFamScan_AllProt", header = F, na.strings = "", sep = "\t")
names(KOprot) = c("protID","Annot")
# Multiple Kegg IDs can be associated to one protein, we need them on one line.
agg_kegg = function(x) {
  keggs = as.character(x$Annot)
  keggs = unique(keggs)
  keggs = keggs[!is.na(keggs)]
  return(paste(keggs, collapse=" "))
}

Kegg_oneline = KOprot %>%
  group_by(protID) %>%
  do(data.frame(Annot_oneline=agg_kegg(.)))
Kegg_oneline = as.data.frame(Kegg_oneline)
Kegg_oneline$Annot_oneline = replace(Kegg_oneline$Annot_oneline, Kegg_oneline$Annot_oneline == "", NA)

cc_annot = merge(x=abund_cc,y=annots_egg,by="protID", all.x = T)
cc_kegg = merge(x=abund_cc,y=Kegg_oneline, by="protID", all.x = T)
cc_annot = cc_annot[,c(2,3)]
cc_kegg = cc_kegg[,c(2,3)]
names(cc_kegg)[2] = "Annot"

##### SIZE #####
cc_size = cc_annot %>%
  group_by(CC_ID) %>%
  do(data.frame(size = nrow(.)))
summary(cc_size)

#### Functional indices ####
func_homscore = function(x) {
  annot = as.character(x$Annot)
  nprot = nrow(x)
  annot = unique(annot)
  annot = annot[!is.na(annot)]
  if(length(annot) == 0) {
    homscore = NA
  } else if (length(annot) == 1) {
    homscore = 1
  } else {
    homscore = 1 - length(annot)/nprot    
  }
  return(homscore)
}

homscore_kegg = cc_kegg %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.kegg =func_homscore(.)))
homscore_kegg = as.data.frame(homscore_kegg)

homscore_eggnog = cc_annot %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.eggnog=func_homscore(.)))
homscore_eggnog = as.data.frame(homscore_eggnog)

#summary(homscore_kegg)
#summary(homscore_eggnog)

func_unkscore = function(x) {
  annot = as.character(x$Annot)
  nprot = nrow(x)
  annot = annot[!is.na(annot)]
  unkscore = (nprot-length(annot))/nprot
  return(unkscore)
}

unkscore_eggnog = cc_annot %>%
  group_by(CC_ID) %>%
  do(data.frame(Unknowns.eggnog=func_unkscore(.)))
unkscore_eggnog = as.data.frame(unkscore_eggnog)

unkscore_kegg = cc_kegg %>%
  group_by(CC_ID) %>%
  do(data.frame(Unknowns.kegg=func_unkscore(.)))
unkscore_kegg = as.data.frame(unkscore_kegg)

#summary(unkscore_eggnog)
#summary(unkscore_kegg)

#### Taxonomical indices ####
mags_taxo = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/Mags_Taxo.csv", header = T, sep = ",", na.strings = c("na"))

abund_cc$MAGID = substr(abund_cc$protID, 1,18)
cc_taxo = merge(abund_cc, mags_taxo, by.x = "MAGID", by.y = "MAG")

##### Domain level #####
func_homscore = function(x) {
  annot = as.character(x$Domain)
  nprot = nrow(x)
  annot = unique(annot)
  annot = annot[!is.na(annot)]
  if(length(annot) == 0) {
    homscore = NA
  } else if (length(annot) == 1) {
    homscore = 1
  } else {
    homscore = 1 - length(annot)/nprot    
  }
  return(homscore)
}

homscore_taxo_Domain = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.taxo.Domain =func_homscore(.)))
homscore_taxo_Domain = as.data.frame(homscore_taxo_Domain)

##### Phylum level #####
func_homscore = function(x) {
  annot = as.character(x$Phylum)
  nprot = nrow(x)
  annot = unique(annot)
  annot = annot[!is.na(annot)]
  if(length(annot) == 0) {
    homscore = NA
  } else if (length(annot) == 1) {
    homscore = 1
  } else {
    homscore = 1 - length(annot)/nprot    
  }
  return(homscore)
}

homscore_taxo_Phylum = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.taxo.Phylum =func_homscore(.)))
homscore_taxo_Phylum = as.data.frame(homscore_taxo_Phylum)

func_unkscore = function(x) {
  annot = as.character(x$Phylum)
  nprot = nrow(x)
  annot = annot[!is.na(annot)]
  unkscore = (nprot-length(annot))/nprot
  return(unkscore)
}

unkscore_taxo_Phylum = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Unknowns.taxo.Phylum=func_unkscore(.)))
unkscore_taxo_Phylum = as.data.frame(unkscore_taxo_Phylum)

##### Class level #####
func_homscore = function(x) {
  annot = as.character(x$Class)
  nprot = nrow(x)
  annot = unique(annot)
  annot = annot[!is.na(annot)]
  if(length(annot) == 0) {
    homscore = NA
  } else if (length(annot) == 1) {
    homscore = 1
  } else {
    homscore = 1 - length(annot)/nprot    
  }
  return(homscore)
}

homscore_taxo_Class = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.taxo.Class=func_homscore(.)))
homscore_taxo_Class = as.data.frame(homscore_taxo_Class)

func_unkscore = function(x) {
  annot = as.character(x$Class)
  nprot = nrow(x)
  annot = annot[!is.na(annot)]
  unkscore = (nprot-length(annot))/nprot
  return(unkscore)
}

unkscore_taxo_Class = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Unknowns.taxo.Class=func_unkscore(.)))
unkscore_taxo_Class = as.data.frame(unkscore_taxo_Class)

##### Order level #####
func_homscore = function(x) {
  annot = as.character(x$Order)
  nprot = nrow(x)
  annot = unique(annot)
  annot = annot[!is.na(annot)]
  if(length(annot) == 0) {
    homscore = NA
  } else if (length(annot) == 1) {
    homscore = 1
  } else {
    homscore = 1 - length(annot)/nprot    
  }
  return(homscore)
}

homscore_taxo_Order = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.taxo.Order =func_homscore(.)))
homscore_taxo_Order = as.data.frame(homscore_taxo_Order)

func_unkscore = function(x) {
  annot = as.character(x$Order)
  nprot = nrow(x)
  annot = annot[!is.na(annot)]
  unkscore = (nprot-length(annot))/nprot
  return(unkscore)
}

unkscore_taxo_Order = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Unknowns.taxo.Order=func_unkscore(.)))
unkscore_taxo_Order = as.data.frame(unkscore_taxo_Order)

##### family level #####
func_homscore = function(x) {
  annot = as.character(x$Family)
  nprot = nrow(x)
  annot = unique(annot)
  annot = annot[!is.na(annot)]
  if(length(annot) == 0) {
    homscore = NA
  } else if (length(annot) == 1) {
    homscore = 1
  } else {
    homscore = 1 - length(annot)/nprot    
  }
  return(homscore)
}

homscore_taxo_Family = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.taxo.Family =func_homscore(.)))
homscore_taxo_Family = as.data.frame(homscore_taxo_Family)

func_unkscore = function(x) {
  annot = as.character(x$Family)
  nprot = nrow(x)
  annot = annot[!is.na(annot)]
  unkscore = (nprot-length(annot))/nprot
  return(unkscore)
}

unkscore_taxo_Family = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Unknowns.taxo.Family=func_unkscore(.)))
unkscore_taxo_Family = as.data.frame(unkscore_taxo_Family)

##### Genus level #####
func_homscore = function(x) {
  annot = as.character(x$Genus)
  nprot = nrow(x)
  annot = unique(annot)
  annot = annot[!is.na(annot)]
  if(length(annot) == 0) {
    homscore = NA
  } else if (length(annot) == 1) {
    homscore = 1
  } else {
    homscore = 1 - length(annot)/nprot    
  }
  return(homscore)
}

homscore_taxo_Genus = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.taxo.Genus =func_homscore(.)))
homscore_taxo_Genus = as.data.frame(homscore_taxo_Genus)

func_unkscore = function(x) {
  annot = as.character(x$Genus)
  nprot = nrow(x)
  annot = annot[!is.na(annot)]
  unkscore = (nprot-length(annot))/nprot
  return(unkscore)
}

unkscore_taxo_Genus = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Unknowns.taxo.Genus=func_unkscore(.)))
unkscore_taxo_Genus = as.data.frame(unkscore_taxo_Genus)

##### MAG level #####
func_homscore = function(x) {
  annot = as.character(x$MAGID)
  nprot = nrow(x)
  annot = unique(annot)
  annot = annot[!is.na(annot)]
  if(length(annot) == 0) {
    homscore = NA
  } else if (length(annot) == 1) {
    homscore = 1
  } else {
    homscore = 1 - length(annot)/nprot    
  }
  return(homscore)
}

homscore_taxo_MAG = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Homogeneity.taxo.MAG =func_homscore(.)))
homscore_taxo_MAG = as.data.frame(homscore_taxo_MAG)

func_unkscore = function(x) {
annot = as.character(x$MAGID)
  nprot = nrow(x)
  annot = annot[!is.na(annot)]
  unkscore = (nprot-length(annot))/nprot
  return(unkscore)
}

unkscore_taxo_MAG = cc_taxo %>%
  group_by(CC_ID) %>%
  do(data.frame(Unknowns.taxo.MAG=func_unkscore(.)))
unkscore_taxo_MAG = as.data.frame(unkscore_taxo_MAG)
#####

# Regroup all levels
cc_stats = Reduce(function(x,y) merge(x,y),list(cc_size, homscore_kegg, homscore_eggnog, homscore_taxo_Phylum, 
                                                homscore_taxo_Class, homscore_taxo_Order, homscore_taxo_Family,
                                                homscore_taxo_Genus, homscore_taxo_MAG, unkscore_eggnog, unkscore_kegg, 
                                                unkscore_taxo_Phylum, unkscore_taxo_Class, unkscore_taxo_Order,
                                                unkscore_taxo_Family, unkscore_taxo_Genus, unkscore_taxo_MAG))

#write.table(cc_stats, "/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Tables_CC_2/CC_stats_with_taxo")

#Empty environment before next section
rm(list=ls())
```

Then, we can link the PFCs to the environment.

```{r}
############### Linking the CC DISTRIBUTION with the ENVIRONMENT #################
#  Emile Faure, 2020

######                           Data Importation                                        ######

# Abundance
abund_cc = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Tables_CC_2/TabAbund_CC_mean_CC2", header = T)
abund_cc = as.data.frame(t(abund_cc))

# Environment
envi_cc=read.table("/home/faure/Documents/PhD/Data_Pangeae/Envi_context_Faure_et_al_2020.txt", sep = "\t", header=T)

climato = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/woa_emile.txt")

# Annotations
cc_annot=read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/CC_annot_8080", header=T)
cc_annot$CC_ID = paste("CC", cc_annot$CC_ID, sep="_")
cc_COGCat=read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/CC_COGCat_8080", header=T)
cc_COGCat$CC_ID =  paste("CC", cc_COGCat$CC_ID, sep="_")
CC2Kegg = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/AnnotKOFAM/CC2Kegg_8080", header = T) # cf KeggAnnotCC.R
KeggToPathway = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/AnnotKOFAM/KO_pathways_mapping_file", header = F, sep = "\t", na.strings="")
names(KeggToPathway) = c("Kegg_ID", "Pathway_ID", "Pathway_Description")

# CC_stats
cc_stats = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Tables_CC_2/CC_stats_with_taxo")

# Taxonomy of MAGs
mags_taxo = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/Mags_Taxo.csv", header = T, sep = ",", na.strings = c("na"))

# Correspondance CC-MAGs
CC_2_Mag = read.table("/home/faure/Documents/PhD/MAGs_Project/Data_tables/S80_C80/Tables_CC_2/TabAbund_CC_NOsum_CC2", header = T)

######                            Table matching                                        #######

# First, let's replace the co-assembly ID by "TARA"
list.coass = list("ANE","ANW","ASW","ASE","ION","IOS","PON","PSE","PSW","RED","MED","SOC")
string.coass = paste(unlist(list.coass), collapse = "|")
row.names(abund_cc) = gsub(string.coass, replacement = "TARA", x = row.names(abund_cc))
# Then let's add a 0 to numbers that aren't 3 long
for (i in c(1:nrow(abund_cc))) {
  if (nchar(row.names(abund_cc)[i])<12) {
    row.names(abund_cc)[i] = paste(unlist(strsplit(row.names(abund_cc)[i], split = "_"))[1], "_0", unlist(strsplit(row.names(abund_cc)[i], split = "_"))[2],
                                   "_", unlist(strsplit(row.names(abund_cc)[i], split = "_"))[3], sep = "")
  }
}
row.names(abund_cc) = gsub("72_100M","072_100M", row.names(abund_cc))
# Finally, let's change meters to SRF and DCM
row.names(abund_cc) = gsub("05M","SRF",row.names(abund_cc))
string.dcm = c("30M|35M|40M|25M|75M|45M|100M|60M|90M|66M|80M|50M|65M|70M|17M")
row.names(abund_cc) = gsub(string.dcm, "DCM", row.names(abund_cc))

# Now we need to order abundcc rows like the ones of envi_cc
abund_cc = abund_cc[order(row.names(abund_cc)),]

######                            Data pre-processing                                        #######

# We can now center, scale, inpute NAs, get rid of highly correlated variables, and near zero variance predictors
envi_cc_mltrans = preProcess(envi_cc, method = c("center","scale","knnImpute","corr","nzv"))

envi_cc_ml = predict(envi_cc_mltrans, envi_cc)
# Down to 51 variables !

# Some CCs have constant null abundances and should be removed for further statistical analysis, all are unknown proteins
#names(abund_cc[, sapply(abund_cc, function(v) var(v, na.rm=TRUE)==0)])
#abund_cc[,grep("CC_115015", names(abund_cc))]
#cc_annot[grep("CC_115015", cc_annot$CC_ID),]
#abund_cc[,grep("CC_154820", names(abund_cc))]
#cc_annot[grep("CC_154820", cc_annot$CC_ID),]
#abund_cc[,grep("CC_163862", names(abund_cc))]
#cc_annot[grep("CC_163862", cc_annot$CC_ID),]
#abund_cc[,grep("CC_187508", names(abund_cc))]
#cc_annot[grep("CC_187508", cc_annot$CC_ID),]
#abund_cc[,grep("CC_38191", names(abund_cc))]
#cc_annot[grep("CC_38191", cc_annot$CC_ID),]

abund_cc = abund_cc[,apply(abund_cc, 2, var, na.rm=TRUE) != 0]

# We still have many CCs with very small variance, that can be problematic when building training sets (if variance = 0 on the training set the model can't work)
abund_cc_nzv = preProcess(abund_cc, method = c("nzv"))
abund_cc_var = data.frame(predict(abund_cc_nzv,abund_cc))

######                            First exploration : unsupervised learning             #######

acp = rda(abund_cc)
summary(acp)
#33% of variance on 2 axes.
plot(acp)

afc = cca(abund_cc)
summary(afc)
#17.8% of variance on 2 axes
plot(afc)

######                            Random forest regressions   #######

#setup parallel backend to use many processors
cores=detectCores()
cl <- makeCluster(cores[1]-5) #not to overload your computer
registerDoParallel(cl)

# Launch the models, note that this is quite a long process, you can skip it by charging directly the output using the last line of this section
rf.cc <- foreach(j=1:228914, .combine=rbind) %dopar% {
  library(caret)
  library(randomForest)
  
  # We create 5 data partitions
  set.seed(1994)
  TrainIndex = createDataPartition(y = abund_cc_var[,j], p = .75, list = F, times = 10)
  
  # Vector for mse decrease :
  rf.fit.mse = c(0)
  rf.bench.mse = c(0)
  rf.mse.decrease = c(0) 
  rf.rsquared = c(0)
  
  # Dataframe for coefs
  imp.env = matrix(nrow = 10, ncol = 51)
  
  for (i in c(1:10)) {
    set.seed(1994)
    rf.model = randomForest(abund_cc_var[TrainIndex[,i],j] ~ ., data = envi_cc_ml[TrainIndex[,i], ], ntree = 500, importance = T)
    imp.env[i,] = rf.model$importance[,2]
    rf.pred = predict(rf.model, envi_cc_ml[-TrainIndex[,i], ])
    rf.fit.mse[i] = mean((abund_cc_var[-TrainIndex[,i],j] - rf.pred)^2) 
    rf.bench.mse[i] = mean((abund_cc_var[-TrainIndex[,i],j] - mean(abund_cc_var[TrainIndex[,i],j]))^2)
    rf.mse.decrease[i] = rf.bench.mse[i] - rf.fit.mse[i]
    rf.rsquared[i] = rf.model$rsq[500]
  }
  
  newrow = c(names(abund_cc_var)[j], mean(rf.fit.mse), mean(rf.bench.mse), mean(rf.mse.decrease), mean(rf.rsquared), colMeans(imp.env))
  newrow
}

rf.cc = as.data.frame(rf.cc)

names(rf.cc) = c("CC_ID","Fit_MSE","Benchmark_MSE","MSE_decrease","Rsquared",names(envi_cc_ml))
rownames(rf.cc) = rf.cc$CC_ID
rf.cc = rf.cc[,-1]
#lapply(rf.cc, class) # ALl factor but we want a numeric
rf.cc[] = lapply(rf.cc, function(x) {
  as.numeric(as.character(x))
})
rf.cc = as.data.frame(rf.cc)

mean(rf.cc$Rsquared) #0.0900 so about 9%
max(rf.cc$Rsquared) #0.907 so about 91%
min(rf.cc$Rsquared) #-0.38

# We can export the random forest results table
#write.table(rf.cc, '/home/faure/Documents/PhD/MAGs_Project/Data_tables/RF_Outputs_ALLvariablesCC')
#rf.cc = read.table('/home/faure/Documents/PhD/MAGs_Project/Data_tables/Random_Forest_Outputs/RF_Outputs_ALLvariablesCC')

##############   Graphical exploration of the results ######

# Rsquared
ggplot() + 
  geom_jitter(aes(x= '', y = rf.cc$Rsquared), alpha = 0.2, colour = 'darkgreen', width = 0.38) +
  geom_boxplot(aes(x= '', y = rf.cc$Rsquared), fill="mediumseagreen", alpha = 0.7) +
  labs(x = '', y = expression(R^{2} ~ "of random forest models")) +
  scale_y_continuous(breaks = c(-0.25,0.0,0.25,0.5,0.75))

# Rank of importance
# All models
rank.rf.imp = as.data.frame(t(apply(as.matrix(rf.cc[,c(5:55)]), 1, rank)))
boxplot = melt(as.matrix(rank.rf.imp))
bpall = ggplot() + geom_boxplot(aes(x = reorder(Var2,-value,FUN=median), y = value), data = boxplot) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, face="bold")) +
  labs(x = "", y = "Rank of importance in random forest regressions)")
bpall
rm(rank.rf.imp, boxplot, bpall)

# Only models with Rsquared > 0
rank.rf.imp.0 = as.data.frame(t(apply(as.matrix(rf.cc[which(rf.cc$Rsquared>0),c(5:55)]), 1, rank)))
rank.rf.imp.0 = 52-rank.rf.imp.0
names(rank.rf.imp.0) = c('Season', "Season moment","Depth","Marine biome","Ocean region","Biogeographical province",
                         "Latitude","Longitude","Max sampling depth","CO3","Alkalinity total","Calcite saturation state",
                         "NO2","Si","NO3","Temperature","Salinity","Oxygen","Chlorophyll A","Opt. backscat. coef. 470nm",   
                         "Fluorescence","Moon phase","Sunshine duration","Iron 5m","Ammonium 5m","NO2 5m", "NO3 5m",
                         "SST gradient", "Okubo Weiss","Lyapunov","Residence time","Euphotic zone depth","Mixed layer depth",
                         "Depth of DCM","Depth max. Brunt Väisälä","Depth Max O2","Depth Min O2","Depth nitracline",
                         "Carbon flux","Bathymetry","Coast Distance","Temperature anomaly","Salinity anomaly","Annual density",
                         "Conductivity anomaly","Dissolved O2 anomaly","Annual percent O2 saturation","Percent O2 saturation anomaly",
                         "Silicate anomaly","Nitrate anomaly","Phosphate anomaly")
boxplot = melt(as.matrix(rank.rf.imp.0))
bp0 = ggplot() + geom_boxplot(aes(x = reorder(Var2,value,FUN=median), y = value), data = boxplot) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, face="bold", size = 12)) +
  labs(x = "", y = "Rank of importance in random forest regressions")
bp0
rm(rank.rf.imp.0, boxplot, bp0)

# Mean Rsquared depending on the most important variable
mean.rsq.per1strankedvar = data.frame(variable = c(0), mean.rsq = c(0), nmodels = c(0))
for(i in c(1:51)) {
  n = names(rank.rf.imp)[i]
  m = mean(rf.cc[which(row.names(rf.cc) %in% row.names(rank.rf.imp[which(rank.rf.imp[,i]==51),])),4])
  nm = nrow(rank.rf.imp[which(rank.rf.imp[,i]==51),])
  mean.rsq.per1strankedvar[i,1] = n
  mean.rsq.per1strankedvar[i,2] = m
  mean.rsq.per1strankedvar[i,3] = nm
}
mean.rsq.per1strankedvar$variable = reorder(mean.rsq.per1strankedvar$variable, -mean.rsq.per1strankedvar$mean.rsq)
ggplot() + geom_col(aes(x=variable,y=mean.rsq, fill=log(nmodels)), data = mean.rsq.per1strankedvar) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, face="bold"))
rm(mean.rsq.per1strankedvar)


##############    Ordination of highest Rsquared CCs ####

# We need to homogeneize the names of CCs among datasets
row.names(rf.cc)[which(row.names(rf.cc)=="CC_1e.05")] = "CC_100000"
row.names(rf.cc)[which(row.names(rf.cc)=="CC_2e.05")] = "CC_200000"
cc_annot$CC_ID[which(cc_annot$CC_ID == "CC_1e+05")] = "CC_100000"
cc_annot$CC_ID[which(cc_annot$CC_ID == "CC_2e+05")] = "CC_200000"
cc_COGCat$CC_ID[which(cc_COGCat$CC_ID == "CC_1e+05")] = "CC_100000"
cc_COGCat$CC_ID[which(cc_COGCat$CC_ID == "CC_2e+05")] = "CC_200000"
names(abund_cc)[which(names(abund_cc)=="CC_1e+05")] = "CC_100000"
names(abund_cc)[which(names(abund_cc)=="CC_2e+05")] = "CC_200000"
names(abund_cc_var)[which(names(abund_cc_var)=="CC_1e.05")] = "CC_100000"
names(abund_cc_var)[which(names(abund_cc_var)=="CC_2e.05")] = "CC_200000"

# We select only PFCs corresponding to models with Rsquared > 0.5
abund_cc_rsq50 = abund_cc_var[,which(names(abund_cc_var) %in% row.names(rf.cc[which(rf.cc$Rsquared>0.50),]))]

# First let's try a CA
afc_50 = cca(abund_cc_rsq50)
#summary(afc_50) #38.9% on two axis
#plot(afc_50) # quite nice distribution ! Let's explore it with a nicer triplot :
# Sites in background
station <- scores(afc_50, scaling=2, choices = c(1:3))$sites
station = as.data.frame(station)
station = merge(station, envi_cc_ml, by = "row.names")
rownames(station) = station[,1]
station = station[,-1]
station = station[,c(1,2,3,9)]
station[,4] = sub(".*\\[(.*)\\].*", "\\1", station[,4], perl=TRUE)
station = within(station,Biogeographical.province <- factor(Biogeographical.province,levels=c("ANTA","FKLD", "SATL","BENG","GUIA","CARB","CAMR","NAST-E","NAST-W","GFST", 
                                                                                              "MEDI","REDS", "ARAB","MONS","EAFR","ISSG", "NPST","PNEC","PEOD","SPSG","CHIL")))
ccscore <- scores(afc_50, scaling=2, choices = c(1:3))$species
ccscore = as.data.frame(ccscore)
gafc <- ggplot() +   geom_hline(yintercept = 0, linetype='dotted') +
  geom_vline(xintercept = 0, linetype='dotted') +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_point(aes(x = ccscore$CA1, y = ccscore$CA2),size = 0.5, alpha = 0.4) +
  geom_point(aes(station[,1], station[,2], col = station[,4]), size = 3.5, alpha = 0.95)+
  scale_color_manual(values = c("#993366","#9966CC", "#66CCFF","#339999","#3399CC","#006699","#0066CC","#003399","#003366","#000099", 
                                "#FFCC33","#CC6633", "#99FF66","#33CC00","#339900","#336600", "#FF9966","#CC6633","#CC0000","#990000","#660000"))  +
  labs(col = "Longhurst province") +
  #  xlab("RDA1 (61.13%)") +  ylab("RDA2 (23.42%)") #rsq75
  xlab("CA1 (21.35%)") +  ylab("CA2 (17.59%)")  #rsq50
# Adding segments for the species scores
gafc

# Now we can go further with a CCA
cca_cc_rsq50 = cca(abund_cc_rsq50 ~ ., data = envi_cc_ml)
#summary(cca_cc_rsq50) # 38.51% on two axis
#RsquareAdj(cca_cc_rsq50) # 75.8%
# Select most informative environmental drivers
set.seed(1994)
cca.step.both = ordistep(cca(abund_cc_rsq50~1,data=envi_cc_ml), scope=formula(cca_cc_rsq50), direction="both", pstep=1000)
#formula(cca.step.both)
# cca.step.both = cca(abund_cc_rsq50 ~ Biogeographical.province + Season.moment + Temperature +
# Depth.euph.zone + Longitude + Depth.bottom.max + Opt.backscat.coef.470nm +
# Depth.Min.O2 + Calcite.saturation.state + Fluorescence + NO3.5m + ChlorophyllA + Ocean.region, data = envi_cc_ml)

res.cca=cca(formula=formula(cca.step.both), data=envi_cc_ml)
summary(res.cca) # 37.64% on two axis
RsquareAdj(res.cca) # 72.9%
#anova.cca(res.cca, by = "axis") 
# CCA1      1   0.35794 69.8736  0.001 ***
# CCA2      1   0.29561 57.7063  0.001 ***
#   CCA3      1   0.19480 38.0266  0.001 ***
#   CCA4      1   0.18239 35.6057  0.001 ***
#   CCA5      1   0.08818 17.2132  0.001 ***
#   CCA6      1   0.05307 10.3596  0.001 ***
#   CCA7      1   0.04347  8.4856  0.001 ***
#   CCA8      1   0.03769  7.3580  0.009 ** 
#   CCA9      1   0.03159  6.1668  0.026 *  

# Triplot
# Sites in background
station <- scores(res.cca, scaling=2, choices = c(1:3))$sites
station = as.data.frame(station)
station = merge(station, envi_cc_ml, by = "row.names")
rownames(station) = station[,1]
station = station[,-1]
station = station[,c(1,2,3,9)]
station[,4] = sub(".*\\[(.*)\\].*", "\\1", station[,4], perl=TRUE)
station = within(station,Biogeographical.province <- factor(Biogeographical.province,levels=c("ANTA","FKLD", "SATL","BENG","GUIA","CARB","CAMR","NAST-E","NAST-W","GFST", 
                                                                                            "MEDI","REDS", "ARAB","MONS","EAFR","ISSG", "NPST","PNEC","PEOD","SPSG","CHIL")))
# CC scores in the background as well
ccscore <- scores(res.cca, scaling=2, choices = c(1:3))$species
ccscore = as.data.frame(ccscore)

g1 <- ggplot() +   geom_hline(yintercept = 0, linetype='dotted') +
  geom_vline(xintercept = 0, linetype='dotted') +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_point(aes(x = ccscore$CCA1, y = ccscore$CCA2),size = 0.5, alpha = 0.4) +
  geom_point(aes(station[,1], station[,2], col = station[,4]), size = 3, alpha = 0.9) +
  scale_color_manual(values = c("#993366","#9966CC", "#66CCFF","#3399CC","#339999","#006699","#0066CC","#003399","#000099","#003366", 
                                "#FFCC33","#FFFF33", "#99FF66","#33CC00","#339900","#336600", "#FF9966","#FF6633","#CC0000","#990000","#660000"))  +
  labs(col = "Longhurst province") +
  xlab("CCA1 (20.62%)") +  ylab("CCA2 (17.03%)")

#scale_color_manual(values = c("#990000","#FF0000","#FFCC33","#FFFF00","#339900","#66CCFF","#003399","#FF00CC")) 

# Adding envi variables 
envi.scores <- scores(res.cca, choices = 1:3, display="bp", scaling=2)[-c(1:30,41),] # we get rid of biogeo provinces
#row.names(envi.scores) = c("Irradiance seasonal anomaly", "Water residence time", "Temperature", "MLD", "Moon phase", "Lyapunov exponent", "NO2")
row.names(envi.scores) = c("Temperature", "Euphotic zone depth", "Longitude", "Max sampling depth", "Optical scattering coefficient 470nm", "O2 minimum depth", "Calcite saturation state", "Fluorescence", "NO3", "Chlorophyll A")

g12 <- g1 + geom_segment(data=as.data.frame(envi.scores),aes(xend = CCA1*4, yend = CCA2*4),x=0,y=0,size = 0.5, linetype="F1",color = 'steelblue4',arrow = arrow(length = unit(0.2,"cm"))) +
  geom_text(aes(envi.scores[,1]*4.2, envi.scores[,2]*4.2, label=rownames(envi.scores)), color="steelblue4")
g12

# And for the 3rd and fourth axis
# Sites in background
station <- scores(res.cca, scaling=2, choices = c(1:4))$sites
station = as.data.frame(station)
station = merge(station, envi_cc_ml, by = "row.names")
rownames(station) = station[,1]
station = station[,-1]
station = station[,c(1,2,3,4,10)]
station[,5] = sub(".*\\[(.*)\\].*", "\\1", station[,5], perl=TRUE)
station = within(station,Biogeographical.province <- factor(Biogeographical.province,levels=c("ANTA","FKLD", "SATL","BENG","GUIA","CARB","CAMR","NAST-E","NAST-W","GFST", 
                                                                                              "MEDI","REDS", "ARAB","MONS","EAFR","ISSG", "NPST","PNEC","PEOD","SPSG","CHIL")))
# CC scores in the background as well
ccscore <- scores(res.cca, scaling=2, choices = c(1:4))$species
ccscore = as.data.frame(ccscore)

g1 <- ggplot() +   geom_hline(yintercept = 0, linetype='dotted') +
  geom_vline(xintercept = 0, linetype='dotted') +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_point(aes(x = ccscore$CCA3, y = ccscore$CCA4),size = 0.5, alpha = 0.4) +
  geom_point(aes(station[,3], station[,4], col = station[,5]), size = 3, alpha = 0.9) +
  scale_color_manual(values = c("#993366","#9966CC", "#66CCFF","#3399CC","#339999","#006699","#0066CC","#003399","#000099","#003366", 
                                "#FFCC33","#FFFF33", "#99FF66","#33CC00","#339900","#336600", "#FF9966","#FF6633","#CC0000","#990000","#660000"))  +
  labs(col = "Longhurst province") +
  xlab("CCA3 (11.22%)") +  ylab("CCA4 (10.51%)")

#scale_color_manual(values = c("#990000","#FF0000","#FFCC33","#FFFF00","#339900","#66CCFF","#003399","#FF00CC")) 

# Adding envi variables 
envi.scores <- scores(res.cca, choices = 1:4, display="bp", scaling=2)[-c(1:30,41),] # we get rid of biogeo provinces
#row.names(envi.scores) = c("Irradiance seasonal anomaly", "Water residence time", "Temperature", "MLD", "Moon phase", "Lyapunov exponent", "NO2")
row.names(envi.scores) = c("Temperature", "Euphotic zone depth", "Longitude", "Max sampling depth", "Optical scattering coefficient 470nm", "O2 minimum depth", "Calcite saturation state", "Fluorescence", "NO3", "Chlorophyll A")

g34 <- g1 + geom_segment(data=as.data.frame(envi.scores),aes(xend = CCA3*4, yend = CCA4*4),x=0,y=0,size = 0.5, linetype="F1",color = 'steelblue4',arrow = arrow(length = unit(0.2,"cm"))) +
  geom_text(aes(envi.scores[,3]*4.2, envi.scores[,4]*4.2, label=rownames(envi.scores)), color="steelblue4")
g34


##############    Exploration of functional annotations among hlePFCs     #########

cc_50_Kegg = CC2Kegg[which(CC2Kegg$CC_ID %in% names(abund_cc_rsq50)),]
#write.table(cc_50_Kegg, "/home/faure/Documents/PhD/MAGs_Project/Data_tables/AnnotKOFAM/CC_rsq50_Kegg", row.names = F, quote = F)

# Which pathways are conserved/not conserved in the 2444
pathway2CC = merge(CC2Kegg, KeggToPathway, by.x = 'KeggID', by.y = 'Kegg_ID')

pathway2CC_rsq50 = pathway2CC[which(pathway2CC$CC_ID %in% names(abund_cc_rsq50)),]
#length(unique(pathway2CC_rsq50$Pathway_Description)) #207

# What pathways are selected
Stats_Pathways = as.data.frame(table(pathway2CC$Pathway_Description))
Stats_Pathways_rsq50 = as.data.frame(table(pathway2CC_rsq50$Pathway_Description))
Stats_Pathways = merge(Stats_Pathways, Stats_Pathways_rsq50, by = "Var1")
rm(Stats_Pathways_rsq50)
names(Stats_Pathways) = c("Pathway_Description", "Freq.tot", "Freq.sel")
Stats_Pathways$perc.pres = Stats_Pathways$Freq.sel / Stats_Pathways$Freq.tot *100
#summary(Stats_Pathways)
Stats_Pathways = Stats_Pathways %>% arrange(desc(perc.pres))
Stats_Pathways[which(Stats_Pathways$Freq.tot>1000),]
ggplot() + geom_col(aes(x=reorder(Pathway_Description, -perc.pres), y = perc.pres), data = Stats_Pathways[which(Stats_Pathways$Freq.tot > 1000 & Stats_Pathways$perc.pres > 1.05),])  +
  geom_hline(yintercept = 1.05, col = "red") +
  theme(axis.text.x = element_text(angle = 90))

# We build a table combining annotations and 
annot_explor = merge(ccscore, cc_50_Kegg, by.x = "row.names", by.y = "CC_ID", all = T)
names(annot_explor)[1] = "CC_ID"
annot_explor = merge(annot_explor, KeggToPathway, by.x = "KeggID", by.y = "Kegg_ID", all.x = T)
annot_explor = annot_explor[order(annot_explor$CC_ID),]

# Detecting pathways with positions different from 0 on each axis
p.value.CCA1 = data.frame(Pathway_Description = c(), p.value = c())
for (i in c(1:length(unique(annot_explor$Pathway_Description)))) {
  p.value.CCA1[i,1] = unique(annot_explor$Pathway_Description)[i]
  if (length(annot_explor$CCA1[which(annot_explor$Pathway_Description==unique(annot_explor$Pathway_Description)[i])]) < 2) {
    p.value.CCA1[i,2] = "NA"
  }
    else {
      p.value.CCA1[i,2] = t.test(annot_explor$CCA1[which(annot_explor$Pathway_Description==unique(annot_explor$Pathway_Description)[i])])$p.value
    }
}
data.CCA1 = annot_explor[which(annot_explor$Pathway_Description %in% p.value.CCA1[which(as.numeric(p.value.CCA1$V2) < 0.00001),1]),]
data.CCA1 = merge(data.CCA1,p.value.CCA1, by.x = "Pathway_Description", by.y = "V1")
ggplot() + geom_boxplot(aes(x=reorder(Pathway_Description,-CCA1,FUN=median), y = CCA1, col = as.numeric(V2)), data = data.CCA1) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, face="bold")) +
  scale_color_continuous(limits = c(1e-100,0.00001), breaks = c(1e-50, 0.00001), guide=guide_colourbar(reverse = TRUE), low="#5EB7F8", high="#1A334B") +
  labs(x = "", y = "Position on CCA1 axis (20.62%)", col = "p-value") +
  ylim(-2,2.5)
rm(data.CCA1, p.value.CCA1)

p.value.CCA2 = data.frame(Pathway_Description = c(), p.value = c())
for (i in c(1:length(unique(annot_explor$Pathway_Description)))) {
  p.value.CCA2[i,1] = unique(annot_explor$Pathway_Description)[i]
  if (length(annot_explor$CCA2[which(annot_explor$Pathway_Description==unique(annot_explor$Pathway_Description)[i])]) < 2) {
    p.value.CCA2[i,2] = "NA"
  }
  else {
    p.value.CCA2[i,2] = t.test(annot_explor$CCA2[which(annot_explor$Pathway_Description==unique(annot_explor$Pathway_Description)[i])])$p.value
  }
}
data.CCA2 = annot_explor[which(annot_explor$Pathway_Description %in% p.value.CCA2[which(as.numeric(p.value.CCA2$V2) < 0.01),1]),]
data.CCA2 = merge(data.CCA2,p.value.CCA2, by.x = "Pathway_Description", by.y = "V1")
ggplot() + geom_boxplot(aes(x=reorder(Pathway_Description,-CCA2,FUN=median), y = CCA2, col = as.numeric(V2)), data = data.CCA2) +
  theme(axis.text.x = element_text(angle = 65, hjust = 1, face="bold", size = 11)) +
  labs(x = "", y = "Position on CCA2 axis (17.03%)", col = "p-value") +
  scale_color_continuous(limits = c(1e-100,0.01), breaks = c(1e-50, 0.001,0.005,0.01), low="#5EB7F8", high="#1A334B", guide = guide_colourbar(reverse=T)) +
  ylim(-1.5,3)
rm(data.CCA2, p.value.CCA2)

# Exploration of unknowns distribution on CCA axes
annot_explor$is.annotated = is.na(annot_explor$Pathway_ID)
a = ggplot() + geom_density(aes(CCA1, fill = is.annotated, col = is.annotated), alpha = 0.4, position = "stack", data = annot_explor) +
  xlim(-2,4) + 
  scale_color_discrete(labels = c("Unknown", "Known")) +
  scale_fill_discrete(labels = c("Unknown", "Known")) +
  labs(y = "Density of connected components", x = "Coordinates along CCA1 (20.62%)", fill = "Kegg pathway\nannotation", col = "Kegg pathway\nannotation")
a

b = ggplot() + geom_density(aes(CCA2, fill = is.annotated, col = is.annotated), alpha = 0.4, position = "stack", data = annot_explor) +
  xlim(-2,3) + 
  coord_flip() +
  scale_color_discrete(labels = c("Unknown", "Known")) +
  scale_fill_discrete(labels = c("Unknown", "Known")) +
  labs(y = "Density of connected components", x = "Coordinates along CCA2 (17.03%)", fill = "Kegg pathway\nannotation", col = "Kegg pathway\nannotation")
b

# Same for eggNOG unknowns:
unkegg_expl = data.frame(CC_ID = row.names(ccscore), CCA1 = ccscore$CCA1, CCA2 = ccscore$CCA2)
unkegg_expl$is.annotated = rep("Known", nrow(unkegg_expl))
unkegg_expl$is.annotated[which(unkegg_expl$CC_ID %in% cc_stats[which(cc_stats$Unknowns.eggnog == 1),1])] = "Unknown"
ggplot() + geom_density(aes(CCA1, fill = is.annotated, col = is.annotated), alpha = 0.4, position = "stack", data = unkegg_expl) +
  xlim(-2,4) + 
  scale_color_discrete(labels = c("Unknown", "Known")) +
  scale_fill_discrete(labels = c("Unknown", "Known")) +
  labs(y = "Density of connected components", x = "Coordinates along CCA1 (20.62%)", fill = "EggNOG annotation", col = "EggNOG annotation")

ggplot() + geom_density(aes(CCA2, fill = is.annotated, col = is.annotated), alpha = 0.4, position = "stack", data = unkegg_expl) +
  xlim(-2,3) + 
  scale_color_discrete(labels = c("Unknown", "Known")) +
  scale_fill_discrete(labels = c("Unknown", "Known")) +
  labs(y = "Density of connected components", x = "Coordinates along CCA2 (17.03%)", fill = "EggNOG annotation", col = "EggNOG annotation")

######## Convex Hulls of KEGGpathways in the CCA space #########

back.convhull = ggplot() + geom_hline(yintercept = 0, linetype='dotted') +
  geom_vline(xintercept = 0, linetype='dotted') +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_point(aes(x=CCA1, y=CCA2), data = annot_explor, size = 0.5, alpha = 0.2)
#path.list = unique(annot_explor$Pathway_Description) # To launch it on all pathways, also adapt col vec
path.list=c("DNA replication", "Fatty acid biosynthesis", "Mismatch repair", 
            "Carotenoid biosynthesis", "Flagellar assembly", "Oxidative phosphorylation",
            "Quorum sensing", "ABC transporters", "Biosynthesis of antibiotics",
            "Carbon fixation pathways in prokaryotes", "Nitrogen metabolism", "Methane metabolism")
graph.list = list()
col.vec = c("dodgerblue4","dodgerblue4","dodgerblue4", "firebrick", "firebrick", "firebrick", "orange3", "orange3", "orange3",
            "springgreen4", "springgreen4", "springgreen4")
for (i in c(1:length(path.list))) {
  focal.path = path.list[i]
  path.chull = annot_explor[which(annot_explor$Pathway_Description == focal.path),c(3:6)]
  hpts12 <- chull(path.chull[,c(1:2)])
  hpts12 <- c(hpts12, hpts12[1])
  g = back.convhull + geom_polygon(aes(x=CCA1, y=CCA2), data = path.chull[hpts12,], fill = col.vec[i], col = col.vec[i], alpha = 0.5) +
    geom_point(aes(x=CCA1, y=CCA2), data = annot_explor[which(annot_explor$Pathway_Description==focal.path),], 
               col = "yellow", size =1.3) +
    geom_point(aes(x=CCA1, y=CCA2), data = annot_explor[which(annot_explor$Pathway_Description==focal.path),], 
               col = "grey10", size =1.3, pch = 1, alpha=0.5) +
    labs(title = path.list[i], x = "CCA1 (20.62%)", y = "CCA2 (17.03%)") +
    theme(text=element_text(size = 7))
  graph.list[[i]] = g
}
# Chull for eggnog unknowns
path.chull = annot_explor[which(annot_explor$CC_ID %in% cc_stats$CC_ID[which(cc_stats$Unknowns.eggnog == 1)]),c(3:6)]
path.chull = distinct(path.chull) # 192 fully unknown CCs in eggNOG annots
hpts12 <- chull(path.chull[,c(1:2)])
hpts12 <- c(hpts12, hpts12[1])
g = back.convhull + geom_polygon(aes(x=CCA1, y=CCA2), data = path.chull[hpts12,], fill = "grey20", col = "grey20", alpha = 0.5) +
  geom_point(aes(x=CCA1, y=CCA2), data = path.chull, 
          col = "yellow", size =1.3) +
  geom_point(aes(x=CCA1, y=CCA2), data = path.chull, 
             col = "grey10", size = 1.3, pch = 1, alpha=0.5) +  
  labs(title = "EggNOG NA", x = "CCA1 (20.62%)", y = "CCA2 (17.03%)") +
  theme(text=element_text(size = 7))
graph.list[[length(graph.list)+1]] = g
# Chull for KEGG unknowns
path.chull = annot_explor[which(annot_explor$CC_ID %in% cc_stats$CC_ID[which(cc_stats$Unknowns.kegg == 1)]),c(3:6)]
path.chull = distinct(path.chull) # 828 fully unknown CCs in eggNOG annots
hpts12 <- chull(path.chull[,c(1:2)])
hpts12 <- c(hpts12, hpts12[1])
g = back.convhull + geom_polygon(aes(x=CCA1, y=CCA2), data = path.chull[hpts12,], fill = "grey20", col = "grey20", alpha = 0.5) +
  geom_point(aes(x=CCA1, y=CCA2), data = path.chull, 
             col = "yellow", size =1.3) +
  geom_point(aes(x=CCA1, y=CCA2), data = path.chull, 
             col = "grey10", size = 1.3, pch = 1, alpha=0.5) +  
  labs(title = "KEGG NA", x = "CCA1 (20.62%)", y = "CCA2 (17.03%)") +
  theme(text=element_text(size = 7))
graph.list[[length(graph.list)+1]] = g

do.call("grid.arrange", c(graph.list,ncol=floor(sqrt(length(graph.list)))))
rm(path.chull, hpts12, focal.path, graph.list, path.list)

# Interesting pathways : Synthesis and degradation of ketone bodies, Fatty acid biosynthesis, Fatty acid metabolism,
# DNA replication, Carotenoid biosynthesis, Flagellar assembly, Pentose and glucuronate interconversions, 
# Quorum sensing, ABC transporters, Biosynthesis of antibiotics, Nitrogen metabolism, Glycerophospholipid metabolism

#####   Centroids of KEGG pathways in the CCA space #####

# To compute one centroid for all pathways
path.list = unique(annot_explor$Pathway_Description)
path.centroid = data.frame(Pathway = path.list, x.coord = rep(0, length(path.list)), y.coord = rep(0,length(path.list)), 
                           x.min= rep(0,length(path.list)), x.max= rep(0,length(path.list)), y.min= rep(0,length(path.list)),
                           y.max= rep(0,length(path.list)), x.plusSD = rep(0,length(path.list)), x.minusSD = rep(0,length(path.list)),
                           y.plusSD = rep(0,length(path.list)), y.minusSD = rep(0,length(path.list)), n.path = rep(0,length(path.list)))

for (i in c(1:length(path.list))) {
  focal.path = path.list[i]
  path.centroid[i,2] = mean(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,3] = mean(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,4] = min(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,5] = max(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,6] = min(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,7] = max(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,8] = path.centroid[i,2] + sd(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,9] = path.centroid[i,2] - sd(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,10] = path.centroid[i,3] + sd(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,11] = path.centroid[i,3] - sd(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,12] = nrow(annot_explor[which(annot_explor$Pathway_Description == focal.path),])
}

# Specific pathways
path.list=c("DNA replication", "Mismatch repair", "Pentose and glucuronate interconversions",
            "Carotenoid biosynthesis", "Oxidative phosphorylation","Flagellar assembly")
path.centroid = data.frame(Pathway = path.list, x.coord = rep(0, length(path.list)), y.coord = rep(0,length(path.list)), 
                           x.min= rep(0,length(path.list)), x.max= rep(0,length(path.list)), y.min= rep(0,length(path.list)),
                           y.max= rep(0,length(path.list)), x.plusSD = rep(0,length(path.list)), x.minusSD = rep(0,length(path.list)),
                           y.plusSD = rep(0,length(path.list)), y.minusSD = rep(0,length(path.list)), n.path = rep(0,length(path.list)),
                           stringsAsFactors = FALSE)
for (i in c(1:length(path.list))) {
  focal.path = path.list[i]
  path.centroid[i,2] = mean(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,3] = mean(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,4] = min(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,5] = max(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,6] = min(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,7] = max(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,8] = path.centroid[i,2] + sd(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,9] = path.centroid[i,2] - sd(annot_explor[which(annot_explor$Pathway_Description == focal.path),3])
  path.centroid[i,10] = path.centroid[i,3] + sd(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,11] = path.centroid[i,3] - sd(annot_explor[which(annot_explor$Pathway_Description == focal.path),4])
  path.centroid[i,12] = nrow(annot_explor[which(annot_explor$Pathway_Description == focal.path),])
}

# Add Dark matter PFCs
cc_stats_rf50 = cc_stats[which(cc_stats$CC_ID %in% cc_50_Kegg$CC_ID),]
focal.darkmatt = as.character(unique(cc_stats_rf50[which(cc_stats_rf50$Unknowns.eggnog == 1 & cc_stats_rf50$Unknowns.kegg == 1 & cc_stats_rf50$Unknowns.taxo.Class == 1),]$CC_ID))
path.centroid[7,] = rep(0, 12)
path.centroid[7,2] = mean(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),3]))
path.centroid[7,3] = mean(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),4]))
path.centroid[7,4] = min(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),3]))
path.centroid[7,5] = max(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),3]))
path.centroid[7,6] = min(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),4]))
path.centroid[7,7] = max(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),4]))
path.centroid[7,8] = path.centroid[7,2] + sd(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),3]))
path.centroid[7,9] = path.centroid[7,2] - sd(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),3]))
path.centroid[7,10] = path.centroid[7,3] + sd(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),4]))
path.centroid[7,11] = path.centroid[7,3] - sd(unique(annot_explor[which(annot_explor$CC_ID %in% focal.darkmatt),4]))
path.centroid[7,12] = length(focal.darkmatt)
path.centroid[7,1] = "Microbial dark matter"

path.centroid = within(path.centroid,Pathway <- factor(Pathway,levels=c("Pentose and glucuronate interconversions","DNA replication", "Mismatch repair",
                                                                        "Carotenoid biosynthesis", "Flagellar assembly", "Oxidative phosphorylation",
                                                                        "Microbial dark matter")))

ggplot() + geom_hline(yintercept = 0, linetype='dotted') +
  geom_vline(xintercept = 0, linetype='dotted') +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_point(aes(x=CCA1, y=CCA2), data = annot_explor, size = 0.5, alpha = 0.2, col = "grey") +
  geom_errorbarh(aes(xmin = x.minusSD, xmax = x.plusSD, y = y.coord, col = Pathway), height = 0.1, data = path.centroid) +
  geom_errorbar(aes(ymin = y.minusSD, ymax = y.plusSD, x = x.coord, col = Pathway), width = 0.1, data = path.centroid) +
  geom_point(aes(x=x.coord, y=y.coord, col = Pathway, size = n.path), data = path.centroid) +
  scale_color_manual(values = c("darkorchid4", "darkorchid1","deeppink2", "orange2", "darkorange1", "darkorange3", "black")) +
  scale_size_continuous(range = c(5,12), breaks = c(5,20,50)) +
  geom_text(aes(x=x.coord, y = y.coord, label = as.character(n.path)), col = "white", data = path.centroid, fontface = "bold", size = 4) +
  labs(x = "CCA1 (20.62%)", y = "CCA2 (17.03%)", size = "Number of associated PFCs", col = "Metabolic pathway barycenter") +
  xlim(-2,2) +
  ylim(-1.5,3)

##########    Exploration of the taxonomy of hlePFCs in the CCA space       #########

CC_2_Mag$Protein_ID = row.names(CC_2_Mag)
CC_2_Mag = CC_2_Mag[,c(1,95)]
CC_2_Mag$Mag_ID = substr(CC_2_Mag$Protein_ID, 1, 18)

CC_2_Mag_Rsq50 = CC_2_Mag[which(CC_2_Mag$CC_ID %in% names(abund_cc_rsq50)),] #13536 Proteins are conserved in the 2444 CCs --> Mean size = 5.54
#max(table(droplevels(CC_2_Mag_Rsq50$CC_ID))) # Max size = 196
#min(table(droplevels(CC_2_Mag_Rsq50$CC_ID))) # Min size = 2
#median(table(droplevels(CC_2_Mag_Rsq50$CC_ID))) # Median = 2

#length(unique(CC_2_Mag$Mag_ID)) #885 MAGs in the CCS
#length(unique(CC_2_Mag_Rsq50$Mag_ID)) #770 Mags represented in the 2444 CCs with more than 50% Rsquare --> 87% of them.

taxo_explor = merge(ccscore, CC_2_Mag_Rsq50, by.x = "row.names", by.y = "CC_ID", all.y = T)
#head(taxo_explor)

#unique(taxo_explor[which(taxo_explor$CCA1<(-1.5)),7]) #Southern Ocean side #19 MAGs in 20 CCs, 5 of which being from the soc assembly (tot mag in soc = 13)
#unique(taxo_explor[which(taxo_explor$CCA1>(0.75)),7]) #Indian Ocean side #47 MAGs in 76 CCs, 12 of which being from ios/ion assemblies (tot mag in ios+ion = 128)
#unique(taxo_explor[which(taxo_explor$CCA2>(1.5)),7]) #Mediterranean side #25 MAGs in 145 CCs, 21 of which being from the med assembly (tot nmag in med = 141)

#mags_taxo[which(mags_taxo$MAG %in% unique(taxo_explor[which(taxo_explor$CCA1<(-1.5)),7])),]
#mags_taxo[which(mags_taxo$MAG %in% unique(taxo_explor[which(taxo_explor$CCA1>(0.75)),7])),]
#mags_taxo[which(mags_taxo$MAG %in% unique(taxo_explor[which(taxo_explor$CCA2>(2)),7])),]

#######   Barycenter of phylums  #####
# First we need to create a taxo_explor table perfectly analogous to annot_explor
taxo_explor = merge(taxo_explor, mags_taxo, by.x = "Mag_ID", by.y = "MAG")
names(taxo_explor)[2] = "CC_ID"

taxo.list=c("Candidatus_Marinimicrobia ", "Spirochaetes", "Verrucomicrobia", "Chloroflexi")
# taxo.list = unique(taxo_explor$Phylum)
# taxo.list = taxo.list[!is.na(taxo.list)]

taxo.centroid = data.frame(Taxon = taxo.list, x.coord = rep(0, length(taxo.list)), y.coord = rep(0,length(taxo.list)), 
                           x.min= rep(0,length(taxo.list)), x.max= rep(0,length(taxo.list)), y.min= rep(0,length(taxo.list)),
                           y.max= rep(0,length(taxo.list)), x.plusSD = rep(0,length(taxo.list)), x.minusSD = rep(0,length(taxo.list)),
                           y.plusSD = rep(0,length(taxo.list)), y.minusSD = rep(0,length(taxo.list)), n.tax = rep(0,length(taxo.list)))

# We add a "unique" to each line compared to the code exploring annot_explor
# because here, each PFC can have multiple rows matching a Phylum
for (i in c(1:length(taxo.list))) {
  focal.taxon = taxo.list[i]
  taxo.centroid[i,2] = mean(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),3]))
  taxo.centroid[i,3] = mean(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),4]))
  taxo.centroid[i,4] = min(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),3]))
  taxo.centroid[i,5] = max(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),3]))
  taxo.centroid[i,6] = min(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),4]))
  taxo.centroid[i,7] = max(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),4]))
  taxo.centroid[i,8] = taxo.centroid[i,2] + sd(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),3]))
  taxo.centroid[i,9] = taxo.centroid[i,2] - sd(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),3]))
  taxo.centroid[i,10] = taxo.centroid[i,3] + sd(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),4]))
  taxo.centroid[i,11] = taxo.centroid[i,3] - sd(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),4]))
  taxo.centroid[i,12] = length(unique(taxo_explor[which(taxo_explor$Phylum == focal.taxon),]$CC_ID))
}

taxo.centroid = within(taxo.centroid,Taxon <- factor(Taxon,levels=c("Candidatus_Marinimicrobia ","Verrucomicrobia",
                                                                    "Chloroflexi","Spirochaetes")))

ggplot() + geom_hline(yintercept = 0, linetype='dotted') +
  geom_vline(xintercept = 0, linetype='dotted') +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_point(aes(x=CCA1, y=CCA2), data = taxo_explor, size = 0.5, alpha = 0.2, col = "grey") +
  geom_errorbarh(aes(xmin = x.minusSD, xmax = x.plusSD, y = y.coord, col = Taxon), height = 0.1, data = taxo.centroid) +
  geom_errorbar(aes(ymin = y.minusSD, ymax = y.plusSD, x = x.coord, col = Taxon), width = 0.1, data = taxo.centroid) +
  geom_point(aes(x=x.coord, y=y.coord, col = Taxon, size = n.tax), data = taxo.centroid) +
  geom_text(aes(x=x.coord, y = y.coord, label = as.character(n.tax)), col = "white", data = taxo.centroid, fontface = "bold", size = 4) +
  labs(x = "CCA1 (20.62%)", y = "CCA2 (17.03%)", size = "Number of associated PFCs", col = "Class barycenter") +
  scale_size_continuous(range = c(6,15), breaks = c(10,40,80,120)) +
  scale_color_manual(values = c("gold3", "darkorange3", "darkolivegreen4", "darkorchid2")) +
  xlim(-2,2) +
  ylim(-1.5,3)

#######   Barycenter of classes  #####

taxo.list=c("Chroococcales", "Prochlorales", "Betaproteobacteria",
            "Opitutae", "Dehalococcoidetes","Spirochaetia")
# taxo.list = unique(taxo_explor$Class)
# taxo.list = taxo.list[!is.na(taxo.list)]

taxo.centroid = data.frame(Taxon = taxo.list, x.coord = rep(0, length(taxo.list)), y.coord = rep(0,length(taxo.list)), 
                           x.min= rep(0,length(taxo.list)), x.max= rep(0,length(taxo.list)), y.min= rep(0,length(taxo.list)),
                           y.max= rep(0,length(taxo.list)), x.plusSD = rep(0,length(taxo.list)), x.minusSD = rep(0,length(taxo.list)),
                           y.plusSD = rep(0,length(taxo.list)), y.minusSD = rep(0,length(taxo.list)), n.tax = rep(0,length(taxo.list)))

# We add a "unique" to each line compared to the code exploring annot_explor
# because here, each PFC can have multiple rows matching a Phylum
for (i in c(1:length(taxo.list))) {
  focal.taxon = taxo.list[i]
  taxo.centroid[i,2] = mean(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),3]))
  taxo.centroid[i,3] = mean(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),4]))
  taxo.centroid[i,4] = min(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),3]))
  taxo.centroid[i,5] = max(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),3]))
  taxo.centroid[i,6] = min(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),4]))
  taxo.centroid[i,7] = max(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),4]))
  taxo.centroid[i,8] = taxo.centroid[i,2] + sd(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),3]))
  taxo.centroid[i,9] = taxo.centroid[i,2] - sd(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),3]))
  taxo.centroid[i,10] = taxo.centroid[i,3] + sd(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),4]))
  taxo.centroid[i,11] = taxo.centroid[i,3] - sd(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),4]))
  taxo.centroid[i,12] = length(unique(taxo_explor[which(taxo_explor$Class == focal.taxon),]$CC_ID))
}

taxo.centroid = within(taxo.centroid,Taxon <- factor(Taxon,levels=c("Betaproteobacteria","Spirochaetia","Opitutae", "Dehalococcoidetes",
                                                                    "Chroococcales", "Prochlorales")))


ggplot() + geom_hline(yintercept = 0, linetype='dotted') +
  geom_vline(xintercept = 0, linetype='dotted') +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_point(aes(x=CCA1, y=CCA2), data = taxo_explor, size = 0.5, alpha = 0.2, col = "grey") +
  geom_errorbarh(aes(xmin = x.minusSD, xmax = x.plusSD, y = y.coord, col = Taxon), height = 0.1, data = taxo.centroid) +
  geom_errorbar(aes(ymin = y.minusSD, ymax = y.plusSD, x = x.coord, col = Taxon), width = 0.1, data = taxo.centroid) +
  geom_point(aes(x=x.coord, y=y.coord, col = Taxon, size = n.tax), data = taxo.centroid) +
  geom_text(aes(x=x.coord, y = y.coord, label = as.character(n.tax)), col = "white", data = taxo.centroid, fontface = "bold", size = 4) +
  labs(x = "CCA1 (20.62%)", y = "CCA2 (17.03%)", size = "Number of associated PFCs", col = "Class barycenter") +
  scale_size_continuous(range = c(6,15), breaks = c(5,50,250,500)) +
  scale_color_manual(values = c("firebrick1", "darkorchid2", "gold3", "darkorange3", "darkolivegreen4", "limegreen")) +
  xlim(-2,2) +
  ylim(-1.5,3)
  
####### Exploration of the distribution of a few unknown CC #######

# Step 1 : find nice candidates CC, with only NA even in eggnog and good CCA position
cc_unk_egg_score = ccscore[which(row.names(ccscore) %in% cc_stats[which(cc_stats$Unknowns.eggnog == 1.0),1]),]
path.chull = cc_unk_egg_score
hpts12 <- chull(path.chull[,c(1:2)])
hpts12 <- c(hpts12, hpts12[1])
back.convhull + geom_polygon(aes(x=CCA1, y=CCA2), data = path.chull[hpts12,],  fill = "grey20", col = "grey20", alpha = 0.5) +
  #    geom_point(aes(x=CCA1, y=CCA2), data = annot_explor[is.na(annot_explor$KeggID),], 
  #               col = "gold2", size =1.5)
  geom_point(aes(x=CCA1, y=CCA2), data = cc_unk_egg_score, 
             col = "yellow", size =1.3) +
  geom_point(aes(x=CCA1, y=CCA2), data = cc_unk_egg_score, 
             col = "grey10", size =1.3, pch = 1, alpha=0.5) +
  labs(title = "NA", x = "CCA1 (20.62%)", y = "CCA2 (17.03%)") +
  theme(text=element_text(size = 7))

# Exploration of polar side
# cc_unk_egg_score[which(cc_unk_egg_score$CCA1<(-1.55)),]
# cc_stats[which(cc_stats$CC_ID %in% rownames(cc_unk_egg_score[which(cc_unk_egg_score$CCA1<(-1.55)),])),]
# cc_annot[which(cc_annot$CC_ID == "CC_210456"),]
# cc_annot[which(cc_annot$CC_ID == "CC_102286"),]
# cc_annot[which(cc_annot$CC_ID == "CC_161812"),]
# rf.cc[which(row.names(rf.cc)=="CC_210456"),]
# rf.cc[which(row.names(rf.cc)=="CC_102286"),]
# rf.cc[which(row.names(rf.cc)=="CC_161812"),]

# Exploration of Indian side
# cc_unk_egg_score[which(cc_unk_egg_score$CCA1>(1.45)),]
# cc_stats[which(cc_stats$CC_ID %in% rownames(cc_unk_egg_score[which(cc_unk_egg_score$CCA1>(1.45)),])),]
# cc_annot[which(cc_annot$CC_ID == "CC_90349"),]
# cc_annot[which(cc_annot$CC_ID == "CC_90580"),]
# cc_annot[which(cc_annot$CC_ID == "CC_90781"),]
# rf.cc[which(row.names(rf.cc)=="CC_90349"),]
# rf.cc[which(row.names(rf.cc)=="CC_90580"),]
# rf.cc[which(row.names(rf.cc)=="CC_90781"),]

a= ggplot() + geom_point(aes(x=envi_cc$Temperature, y=abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_102286")]), 
                         col = "hotpink1", alpha = 0.8, size = 2) +
  labs(x="Temperature (°C)", y = "Norm. sequence abundance", title = "PFC #102286 (CCA1 = -1.70) ")
b= ggplot() + geom_point(aes(x=envi_cc$Temperature, y=abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_210456")]), 
                      col = "darkorchid1", alpha = 0.8, size = 2) +
  labs(x="Temperature (°C)", y = "Norm. sequence abundance", title = "PFC #210456 (CCA1 = -1.69) ")
c= ggplot() +  geom_point(aes(x=envi_cc$Temperature, y=abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_161812")]), 
             col = "darkmagenta", alpha = 0.8, size = 2) +
  labs(x="Temperature (°C)", y = "Norm. sequence abundance", title = "PFC #161812 (CCA1 = -1.63) ")

d= ggplot() +  geom_point(aes(x=envi_cc$Temperature, y=abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_90580")]), 
                          col = "darkolivegreen2", alpha = 0.8, size = 2) +
  labs(x="Temperature (°C)", y = "Norm. sequence abundance", title = "PFC #90580 (CCA1 = 1.51) ")
e= ggplot() +  geom_point(aes(x=envi_cc$Temperature, y=abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_90349")]), 
             col = "darkolivegreen3", alpha = 0.8, size = 2) +
  labs(x="Temperature (°C)", y = "Norm. sequence abundance", title = "PFC #90349 (CCA1 = 1.47) ")
f= ggplot() +  geom_point(aes(x=envi_cc$Temperature, y=abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_90781")]), 
                          col = "darkolivegreen4", alpha = 0.8, size = 2) +
  labs(x="Temperature (°C)", y = "Norm. sequence abundance", title = "PFC #90781 (CCA1 = 1.46) ")

multiplot(a,b,c,d,e,f, cols = 2)

###   Add 3 examples of darkmatter PFCs associated with mediterranean side

#ccscore[which(row.names(ccscore) %in% focal.darkmatt),]
# CC_172160 1.8256692
# CC_176586 1.6667614
# CC_177371 1.6591272

envi_cc$Med_NoMed = rep("Not Mediterranean", nrow(envi_cc))
envi_cc[which(envi_cc$Biogeographical.province == "[MEDI] Mediterranean Sea, Black Sea Province (MRGID:21465)"),]$Med_NoMed = "Mediterranean"
g = ggplot() + geom_boxplot(aes(x=envi_cc$Med_NoMed, y= abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_172160")],
                                fill = envi_cc$Med_NoMed)) +
  scale_fill_manual(values = c("Gold1", "Grey30")) +
  theme(legend.position = "none", axis.text.x = element_text(size = 11)) +
  labs(x=NULL, y = "Norm. sequence abundance", title = "PFC #172160 (CCA2 = 1.83)")

h = ggplot() + geom_boxplot(aes(x=envi_cc$Med_NoMed, y= abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_176586")],
                                fill = envi_cc$Med_NoMed)) +
  scale_fill_manual(values = c("Gold2", "Grey30")) +
  theme(legend.position = "none", axis.text.x = element_text(size = 11)) +
  labs(x=NULL, y = "Norm. sequence abundance", title = "PFC #176586 (CCA2 = 1.67)")

i = ggplot() + geom_boxplot(aes(x=envi_cc$Med_NoMed, y= abund_cc_rsq50[,which(names(abund_cc_rsq50) == "CC_177371")],
                                fill = envi_cc$Med_NoMed)) +
  scale_fill_manual(values = c("Gold3", "Grey30")) +
  theme(legend.position = "none", axis.text.x = element_text(size = 11)) +
  labs(x=NULL, y = "Norm. sequence abundance", title = "PFC #177371 (CCA2 = 1.66)")

multiplot(a,b,c,g,h,i, cols = 2)

#########   Testing for the Assembly effect  #########

taxo_explor$Assembly = substr(taxo_explor$Mag_ID,6,8)

taxo.list=c("MED","ASE","RED","IOS", "ANW")
#taxo.list = taxo.list[!is.na(taxo.list)]
taxo.centroid = data.frame(Taxon = taxo.list, x.coord = rep(0, length(taxo.list)), y.coord = rep(0,length(taxo.list)), 
                           x.min= rep(0,length(taxo.list)), x.max= rep(0,length(taxo.list)), y.min= rep(0,length(taxo.list)),
                           y.max= rep(0,length(taxo.list)), x.plusSD = rep(0,length(taxo.list)), x.minusSD = rep(0,length(taxo.list)),
                           y.plusSD = rep(0,length(taxo.list)), y.minusSD = rep(0,length(taxo.list)), n.tax = rep(0,length(taxo.list)))
for (i in c(1:length(taxo.list))) {
  focal.taxon = taxo.list[i]
  taxo.centroid[i,2] = mean(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),3]))
  taxo.centroid[i,3] = mean(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),4]))
  taxo.centroid[i,4] = min(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),3]))
  taxo.centroid[i,5] = max(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),3]))
  taxo.centroid[i,6] = min(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),4]))
  taxo.centroid[i,7] = max(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),4]))
  taxo.centroid[i,8] = taxo.centroid[i,2] + sd(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),3]))
  taxo.centroid[i,9] = taxo.centroid[i,2] - sd(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),3]))
  taxo.centroid[i,10] = taxo.centroid[i,3] + sd(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),4]))
  taxo.centroid[i,11] = taxo.centroid[i,3] - sd(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),4]))
  taxo.centroid[i,12] = length(unique(taxo_explor[which(taxo_explor$Assembly == focal.taxon),]$CC_ID))
}


taxo.centroid = within(taxo.centroid,Taxon <- factor(Taxon,levels=c("ASE","ANW","MED", "RED",
                                                                        "IOS")))

ggplot() + geom_hline(yintercept = 0, linetype='dotted') +
  geom_vline(xintercept = 0, linetype='dotted') +
  theme(plot.title=element_text(hjust=0.5)) +
  geom_point(aes(x=CCA1, y=CCA2), data = taxo_explor, size = 0.5, alpha = 0.2, col = "grey") +
  geom_errorbarh(aes(xmin = x.minusSD, xmax = x.plusSD, y = y.coord, col = Taxon), height = 0.1, data = taxo.centroid) +
  geom_errorbar(aes(ymin = y.minusSD, ymax = y.plusSD, x = x.coord, col = Taxon), width = 0.1, data = taxo.centroid) +
  geom_point(aes(x=x.coord, y=y.coord, col = Taxon, size = n.tax), data = taxo.centroid) +
  geom_text(aes(x=x.coord, y = y.coord, label = as.character(n.tax)), col = "white", data = taxo.centroid, fontface = "bold", size = 4) +
  labs(x = "CCA1 (20.62%)", y = "CCA2 (17.03%)", size = "Number of associated PFCs", col = "Assembly barycenter") +
  scale_size_continuous(range = c(10,15), breaks = c(100,500,1000)) +
  scale_color_manual(values=c("darkblue", "lightblue", "#FFCC33", "firebrick", "darkgreen")) +
  xlim(-2,2) +
  ylim(-1.5,3)

```









